src/model/timestamp.rs

```diff
@@ -0,0 +1,184 @@
+        Self(OffsetDateTime::from_unix_timestamp_nanos(ns).expect("can't fail"))
+    }
+
+    /// Create a new `Timestamp` with the current date and time in UTC.
+    pub fn now() -> Self {
+        Self(OffsetDateTime::now_utc())
+    }
+
+    /// Returns the number of non-leap seconds since January 1, 1970 0:00:00 UTC
+    pub fn unix_timestamp(&self) -> i64 {
+        self.0.unix_timestamp()
+    }
+
+    fn parse(input: &str) -> Result<Timestamp, ParseError> {
+        OffsetDateTime::parse(input, &Rfc3339).map(Self).map_err(|_| ParseError)
+    }
+}
+
+#[cfg(not(feature = "time"))]
+impl Timestamp {
+    pub(crate) fn from_discord_id(id: u64) -> Timestamp {
+        Self(Utc.timestamp_millis(((id >> 22) + DISCORD_EPOCH) as i64))
+    }
+
+    /// Create a new `Timestamp` with the current date and time in UTC.
+    pub fn now() -> Self {
+        Self(Utc::now())
+    }
+
+    /// Returns the number of non-leap seconds since January 1, 1970 0:00:00 UTC
+    pub fn unix_timestamp(&self) -> i64 {
+        self.0.timestamp()
+    }
+
+    fn parse(input: &str) -> Result<Timestamp, ParseError> {
+        DateTime::parse_from_rfc3339(input)
+            .map(|d| Self(d.with_timezone(&Utc)))
+            .map_err(|_| ParseError)
+    }
+}
+
+#[derive(Debug)]
+pub struct ParseError;
+
+impl std::error::Error for ParseError {}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("invalid rfc3339 string")
+    }
+}
+
+#[cfg(feature = "time")]
+impl fmt::Display for Timestamp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = self.0.format(&Rfc3339).map_err(|_| fmt::Error)?;
+        f.write_str(&s)
+    }
+}
+
+#[cfg(not(feature = "time"))]
+impl fmt::Display for Timestamp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use chrono::SecondsFormat;
+        let s = self.0.to_rfc3339_opts(SecondsFormat::Millis, true);
+        f.write_str(&s)
+    }
+}
+
+#[cfg(not(feature = "time"))]
+impl Deref for Timestamp {
+    type Target = DateTime<Utc>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[cfg(feature = "time")]
+impl Deref for Timestamp {
+    type Target = OffsetDateTime;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<String> for Timestamp {
+    /// Parses an RFC 3339 date and time string such as `2016-04-30T11:18:25.796Z`.
+    ///
+    /// Panics on invalid value.
+    fn from(s: String) -> Self {
+        #[allow(clippy::unwrap_used)]
+        Timestamp::parse(&s).unwrap()
+    }
+}
+
+impl<'a> From<&'a str> for Timestamp {
+    /// Parses an RFC 3339 date and time string such as `2016-04-30T11:18:25.796Z`.
+    ///
+    /// Panics on invalid value.
+    fn from(s: &'a str) -> Self {
+        #[allow(clippy::unwrap_used)]
+        Timestamp::parse(s).unwrap()
+    }
+}
```

Comment:
Could also offer `TryFrom` variants of these that do not panic upon a `ParseError`. It would make said error type more useful. Currently, it just sits being present in the documentation, but it's nowhere used (i.e. through an interface exposed to the user).

Reply:
i didn't really think of adding `TryFrom` because of how the current timestamp type is used in the builders (`Into<Timestamp>`).

Reply:
I see, but it'd be nice to add those. Otherwise, why make `ParseError` public?

Reply:
I didn't make the `timestamp` module public.

Reply:
Ah, you are right. I was only focusing in the scope of this module. You only re-export the `Timestamp` struct from `model/mod.rs`, so users never actually see `ParseError`. In that case, you can ignore this.

Reply:
I tried to add `TryFrom`. I've seen this error before.
```
error[E0119]: conflicting implementations of trait `std::convert::TryFrom<std::string::String>` for type `model::timestamp::Timestamp`
   --> src/model/timestamp.rs:157:1
    |
157 | impl TryFrom<String> for Timestamp {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `core`:
            - impl<T, U> TryFrom<U> for T
              where U: Into<T>;
```